rust_snuba/src/processors/replays.rs

```diff
@@ -0,0 +1,573 @@
+use anyhow::Context;
+use rust_arroyo::backends::kafka::types::KafkaPayload;
+use serde::{Deserialize, Serialize};
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use uuid::Uuid;
+
+use crate::types::{InsertBatch, KafkaMessageMetadata, RowData};
+
+pub fn process_message(
+    payload: KafkaPayload,
+    metadata: KafkaMessageMetadata,
+) -> anyhow::Result<InsertBatch> {
+    let payload_bytes = payload.payload().context("Expected payload")?;
+
+    let replay_message: ReplayMessage = serde_json::from_slice(payload_bytes)?;
+    let replay_payload = serde_json::from_slice(&replay_message.payload)?;
+
+    let output = match replay_payload {
+        ReplayPayload::ClickEvent(event) => {
+            let replay_rows = event.clicks.into_iter().map(|click| ReplayRow {
```

Comment:
I believe you can do it in one go:
```suggestion
            event.clicks.into_iter().map(|click| serde_json::to_vec(&ReplayRow {
            })).collect()?
```
Its possible Rust will tell you to add a type annotation somewhere though.

Reply:
I don't think this works because the `?` needs to happen within the map's closure which doesn't support error propagation.

Reply:
Calling `collect()` on an iterator that yields `Result` will give you a `Result<Vec<T>, E>` which you unwrap at the very end.

Reply:
The type signature when I try to implement it is `Vec<Result<T>, E>>` which can't be unwrapped.

Reply:
just tried it locally. you can do it with `.collect::<Result<_, _>>()?` -- rust permits both and can't figure out which one you mean

Reply:
That worked.